terraform/s3/main.tf

```diff
@@ -0,0 +1,134 @@
+terraform {
+  required_version = ">= 1.0"
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 5.0"
+    }
+  }
+}
+
+# S3 Bucket
+# Creates the main S3 bucket resource
+resource "aws_s3_bucket" "main" {
+  bucket        = var.bucket_name
+  force_destroy = var.force_destroy
+
+  tags = merge(
+    var.tags,
+    {
+      Name = var.bucket_name
+    }
+  )
+}
+
+# Bucket Versioning
+# Enables versioning to protect against accidental deletion and provide object history
+resource "aws_s3_bucket_versioning" "main" {
+  bucket = aws_s3_bucket.main.id
+
+  versioning_configuration {
+    status = var.versioning_enabled ? "Enabled" : "Suspended"
+  }
+}
+
+# Server-Side Encryption
+# Encrypts all objects stored in the bucket using either SSE-S3 or SSE-KMS
+resource "aws_s3_bucket_server_side_encryption_configuration" "main" {
+  bucket = aws_s3_bucket.main.id
+
+  rule {
+    apply_server_side_encryption_by_default {
+      sse_algorithm     = var.kms_key_id != null ? "aws:kms" : "AES256"
```

Comment:
`sse_algorithm` can be one of `aws:kms`, `AES256`, and `aws:kms:dsse`.
For this time, we only need to support `aws:kms` and `AES256`, so the `aws:kms:dsse` option can be ignored.

reference: [aws_s3_bucket_server_side_encryption_configuration](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket_server_side_encryption_configuration)